--- /root/tmp/sphinx-2.0.4-release/mysqlse/snippets_udf.cc	2012-01-31 01:07:35.000000000 +0200
+++ snippets_udf.cc	2012-03-19 20:04:25.340224806 +0200
@@ -22,8 +22,11 @@
 
 #include <mysql_version.h>
 
-#if MYSQL_VERSION_ID>50100
-#include "mysql_priv.h"
+#if MYSQL_VERSION_ID>=50515
+#include "sql_class.h"
+#include "sql_array.h"
+#elif MYSQL_VERSION_ID>50100
+#include <mysql_priv.h>
 #include <mysql/plugin.h>
 #else
 #include "../mysql_priv.h"
@@ -158,7 +161,7 @@
 
 	SEARCHD_COMMAND_EXCERPT		= 1,
 
-	VER_COMMAND_EXCERPT		= 0x103,
+	VER_COMMAND_EXCERPT		= 0x104,
 };
 
 /// known answers
@@ -380,23 +383,42 @@
 		else
 		{
 			int tmp_errno;
+                        bool bError = false;
+#if MYSQL_VERSION_ID>=50515
+                        struct addrinfo *hp = NULL;
+                        tmp_errno = getaddrinfo ( m_sHost, NULL, NULL, &hp );
+                        if ( !tmp_errno || !hp || !hp->ai_addr )
+                        {
+                                bError = true;
+                                if ( hp )
+                                        freeaddrinfo ( hp );
+                        }
+#else
 			struct hostent tmp_hostent, *hp;
 			char buff2 [ GETHOSTBYNAME_BUFF_SIZE ];
-
 			hp = my_gethostbyname_r ( m_sHost, &tmp_hostent, buff2, sizeof(buff2), &tmp_errno );
 			if ( !hp )
 			{
 				my_gethostbyname_r_free();
+                                bError = true;
+			}
+#endif
+                        if ( bError )
+                        {
+                                char sError[256];
+                                my_snprintf ( sError, sizeof(sError), "failed to resolve searchd host (name=%s)", m_sHost );
 
-				char sError[256];
-				snprintf ( sError, sizeof(sError), "failed to resolve searchd host (name=%s)", m_sHost );
-
-				my_error ( ER_CONNECT_TO_FOREIGN_DATA_SOURCE, MYF(0), sError );
+                                my_error ( ER_CONNECT_TO_FOREIGN_DATA_SOURCE, MYF(0), sError );
 				return -1;
-			}
+                        }
 
-			memcpy ( &sin.sin_addr, hp->h_addr, Min ( sizeof(sin.sin_addr), (size_t)hp->h_length ) );
-			my_gethostbyname_r_free();
+#if MYSQL_VERSION_ID>=50515
+                        memcpy ( &sin.sin_addr, hp->ai_addr, Min ( sizeof(sin.sin_addr), (size_t)hp->ai_addrlen ) );
+                        freeaddrinfo ( hp );
+#else
+                        memcpy ( &sin.sin_addr, hp->h_addr, Min ( sizeof(sin.sin_addr), (size_t)hp->h_length ) );
+                        my_gethostbyname_r_free();
+#endif
 		}
 	} else
 	{
